openvdb/tools/FastSweeping.h

```diff
@@ -0,0 +1,1432 @@
+///
+/// @note    Solves the (simplified) Eikonal Eq: @f$|\nabla \phi|^2 = 1@f$ and
+///          performs velocity extension,  @f$\nable f\nabla \phi = 0@f$, both
+///          by means of the fast sweeping algorithm detailed in:
+///          "A Fast Sweeping Method For Eikonal Equations"
+///          by H. Zhao, Mathematics of Computation, Vol 74(230), pp 603-627, 2004
+///
+/// @details The algorithm used below for parallel fast sweeping was first publised in:
+///          "New Algorithm for Sparse and Parallel Fast Sweeping: Efficient
+///          Computation of Sparse Distance Fields" by K. Museth, ACM SIGGRAPH Talk,
+///          2017, http://www.museth.org/Ken/Publications_files/Museth_SIG17.pdf
+
+#ifndef OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+#define OPENVDB_TOOLS_FASTSWEEPING_HAS_BEEN_INCLUDED
+
+//#define BENCHMARK_FAST_SWEEPING
+
+#include <type_traits>// for static_assert
+#include <cmath>
+#include <limits>
+#include <thread>
+
+#include <tbb/blocked_range.h>
+#include <tbb/parallel_for.h>
+#include <tbb/parallel_sort.h>
+#include <tbb/parallel_invoke.h>
+#include <tbb/enumerable_thread_specific.h>
+
+#include <openvdb/math/Math.h> // for Abs() and isExactlyEqual()
+#include <openvdb/math/Stencils.h> // for GradStencil
+#include <openvdb/util/PagedArray.h>
+#include <openvdb/tree/LeafManager.h>
+#include "LevelSetUtil.h"
+#include "Morphology.h"
+
+#include "Statistics.h"
+#ifdef BENCHMARK_FAST_SWEEPING
+#include <openvdb/util/CpuTimer.h>
+#endif
+
+#include <openvdb/math/Stats.h>
+
+namespace openvdb {
+OPENVDB_USE_VERSION_NAMESPACE
+namespace OPENVDB_VERSION_NAME {
+namespace tools {
+
+/// @brief Converts a scalar fog volume into a signed distance function. Active input voxels
+///        with scalar values above the given isoValue will have NEGATIVE distance
+///        values on output, i.e. they are assumed to be INSIDE the iso-surface.
+///
+/// @return A shared pointer to a signed-distance field defined on the active values
+///         of the input fog volume.
+///
+/// @param fogGrid  Scalar (floating-point) volume from which an
+///                 iso-surface can be defined.
+///
+/// @param isoValue A value which defines a smooth iso-surface that
+///                 intersects active voxels in @a fogGrid.
+///
+/// @param nIter    Number of iterations of the fast sweeping algorithm.
+///                 each performing 2^3 = 8 sweeps.
+///
+/// @note Strictly speaking a fog volume is normalized to the range [0,1] but this
+///       method accepts a scalar volume with an arbitary range, as long as the it
+///       includes the @a isoValue.
+///
+/// @details Topology of output grid is identical to that of the input grid, except
+///          active tiles in the input grid will be converted to active voxels
+///          in the output grid!
+///
+/// @throw  RuntimeError if the iso-surface does not intersected any active
+///         voxels or if it intersects any active tiles in @a fogGrid.
```

Comment:
....  I can see why this is the case, but it reminds me of the problems with volume blur.  Both can be fixed in the same manner, so I don't think this should delay this PR at all; but it does add more urgency to getting a system to densify boundaries.

Reply:
agree on both points